File: server/lib/pdfParser.js
const { PDFParse } = require("pdf-parse");

async function extractText(fileBuffer, mimetype) {
  if (mimetype === "application/pdf") {
    try {
      const pdf = new PDFParse(new Uint8Array(fileBuffer), { verbosity: 0 });
      const result = await pdf.getText();
      return result.pages.map((p) => p.text).join("\n");
    } catch (err) {
      throw new Error("PDF parsing failed: " + err.message);
    }
  }
  if (mimetype === "text/plain") {
    return fileBuffer.toString("utf-8");
  }
  throw new Error("Unsupported file type. Upload PDF or TXT.");
}

module.exports = { extractText };

File: server/lib/pdf_parser.py
import pdfplumber
import io


async def extract_text(file_buffer: bytes, mimetype: str) -> str:
    if mimetype == "application/pdf":
        try:
            with pdfplumber.open(io.BytesIO(file_buffer)) as pdf:
                pages = [page.extract_text() or "" for page in pdf.pages]
                return "\n".join(pages)
        except Exception as err:
            raise Exception(f"PDF parsing failed: {err}")
    if mimetype == "text/plain":
        return file_buffer.decode("utf-8")
    raise Exception("Unsupported file type. Upload PDF or TXT.")
